### PR TITLE
Allow setting Letter case and Decoration to 'None' and add Letter case to Global Styles

### LINKED ISSUE #36735: TextTransformControl/TextDecorationControl: Balancing usability and accessibility
There has been some complications in finding the correct implementation for the `TextTransformControl` and `TextDecorationControl` components (e.g. #26060 #31623). It seems like the difficulty is in making the implementation semantically correct, while also being intuitively usable for sighted users.

Text Transform and Text Decoration controls

## Considered implementations

|  | Pros | Cons |
|---|---|---|
| Radio (ToggleGroupControl) | Mutually exclusive | Cannot re-click to reset (or requires a separate "default" option) |
| Group of buttons | Can click to reset | Not mutually exclusive |
| Rely on ToolsPanel reset | Can use Radio | Not intuitive |

The options items are mutually exclusive, which makes them similar to a radio control. But we also want an intuitive way to reset the control, which is currently achieved by clicking the active item again. We can't implement this behavior onto a radio control because radio buttons are expected to be idempotent.

On the other hand, a mere array of buttons, even if `role="group"`ed, fails to semantically capture the mutual exclusivity of the options.

Any ideas or thoughts? (cc @ciampo @diegohaz)

### PR DESCRIPTION
**Pull request by noisysocks**
## What?

Alternative to https://github.com/WordPress/gutenberg/pull/43356. Partially reverts https://github.com/WordPress/gutenberg/issues/36735. Closes https://github.com/WordPress/gutenberg/issues/36735 and addresses one of the items in https://github.com/WordPress/gutenberg/issues/34345#issuecomment-1217390408.

This PR does three things:

- Go back to using buttons in `TextTransformControl` and `TextDecorationControl` instead of a `ToggleGroupControl`
- Add _None_ option to `TextTransformControl` and `TextDecorationControl` which allows user to set `text-transform: none` and `text-decoration: none`
- Add `TextTransformControl` to Global Styles

### Why?

We want to add `TextTransformControl` to Global Styles. See https://github.com/WordPress/gutenberg/issues/34345#issuecomment-1217479784.

We also want to allow the user to set `text-transform: none` or `text-decoration: none` explicitly. This allows users to override theme styles at a block level. See https://github.com/WordPress/gutenberg/issues/42766 and https://github.com/WordPress/gutenberg/pull/43356#issuecomment-1241657333.

So, using `TextTransformControl` as an example, the control should support these options:

- Default (no style)
- None (`text-transform: none`)
- Upper case (`text-transform: uppercase`)
- Lower case (`text-transform: lowercase`)
- Capitalize (`text-transform: capitalize`)

Displaying all five options is overwhelming. It is much more intuitive for users if we can display only four buttons (_None_, _Upper case_, _Lower case_ and _Capitalize_) and implement _Default_ using a click-to-unselect mechanism or the ellipsis menu in the Tools Panel. Unfortunately though, `ToggleGroupControl` does not support click-to-unselect as it uses `radio`, and Global Styles does not use Tools Panel.

We switched from a group of buttons to a `ToggleGroupControl` in https://github.com/WordPress/gutenberg/issues/36735 for two reasons:

1. So that `TextTransformControl` and `TextDecorationControl` would have updated 40px styling.
2. To communicate semantically using `role=radio` that these options are mutually exclusive.

(1) can be implemented by styling the group of buttons.

I've been reading about (2) and I don't believe it's actually a big concern. There is little practical difference in terms of what is announced by a screen reader between a group of buttons with `aria-pressed` and a group of `radio`s. Moreover, we should use the markup and semantic roles that most closely resembles the UI. In this case, I believe that `TextTransformControl` and `TextDecorationControl` **is** a group of buttons, because:

- They look more like buttons than radios. (They're squares containing an icon, not unfilled circles with an adjacent label)
- They support button-like behaviour such as deselection 
- They aren't grouped together and therefore I wouldn't expect them to be in the same tab stop (i.e. <kbd>Tab</kbd> should move focus between the buttons)
 - `TextDecorationControl` could in theory support multiple values e.g. `text-decoration: underline strike-through`

https://lea.verou.me/2022/07/button-group/ is a good article that I found on this subject.

## Testing Instructions
1. Go to Editor → Global Styles → Typography.
2. Check that _Letter case_ appears in the Heading section but no other sections, and that modifying it works.
2. Go to Global Styles → Blocks.
3. Check that _Letter case_ appears in the Typography panel for blocks that support these controls, and that modifying them works.
4. `npm run storybook:dev` and see the stories for `TextTransformControl`/`TextDecorationControl`.

## Screenshots or screencast 
https://user-images.githubusercontent.com/612155/189600028-ef8ba1e5-7be4-429a-b7e3-b3e242ecbdbd.mp4

### CODE REVIEW

**Comment by jasmussen** [comment]
Thank you for your patience on this one. The behavior in this PR feels correct: there's an explicit "none" added to the controls, you can toggle and untoggle (so you can go back to unset unlike the radio behavior). 

We have a visual challenge, though. Each dark toggle button should be exactly 32x32px, but they are 36px tall and flexing in their width:

Screenshot 2022-09-12 at 10 49 32

If I do a little inspector hacking I can get them to the right sizes, which introduces a different problem, that there isn't room for 3 + 4 buttons:

Screenshot 2022-09-12 at 10 52 07

A few solutions to this one:

* Smaller than 32px icons. Perhaps 24x24? Ideally we can avoid this one as it would mess with the general grid spacing.
* Allow the text decoration control to fill a full row. This might be the interim solution.

Another option is to allow the two columns to be flexible. Mockup:

Screenshot 2022-09-12 at 10 52 07

CC: @javierarce @jameskoster in case you have thoughts, and @ciampo @mirka for thoughts from the component side.

The near term solution, probably, is to allow the text decoration to fill a full row, and to get the buttons to be the correct 32x32px sizes. Does that sound okay?

**Comment by jasmussen** [comment]
Just connecting some dots here, this PR seems to overlap: https://github.com/WordPress/gutenberg/pull/43993 CC: @draganescu @scruffian

**Comment by draganescu** [comment]
I'll absolutely close https://github.com/WordPress/gutenberg/pull/43993 for this one here which is a complete implementation since being able to un-select is essential to preserve theme styles as they are w/ no user intervention.

**Comment by jameskoster** [comment]
Personally I find 24x24 buttons awkward to click, especially on mobile. And the difference in height between buttons / inputs creates unnecessary tension in the overall appearance. Since we're committed to making all inputs 40px tall, I would be inclined to do the same with icon buttons.

If that means some controls (like this one) occupy a single row, then I think it's worth it. In the future we could consider making the Inspector wider to better accommodate multi-column layouts.

**Comment by jasmussen** [comment]
> Since we're committed to making all inputs 40px tall, I would be inclined to do the same with icon buttons.

Just to clarify here, I don't think that means the 32x32px buttons must become 40x40 — or that the toggle control has to be 40px tall suddenly. The buttons can remain 32x32, but inside a 40px wrapper, so the spacing remains right.

**Comment by jameskoster** [comment]
That's the reason I think it's worth considering making the icons 40x40. It gives us this:

Screenshot 2022-09-12 at 14 48 13

Which feels more balanced than:

Screenshot 2022-09-12 at 14 50 14

**Comment by mirka** [comment]
> https://lea.verou.me/2022/07/button-group/ is a good article that I found on this subject.

This is a great pull, thanks! It clears up the HTML concerns, which feels like a big step forward.

However I feel like this solution doesn't fully address one of the main takeaways of the article (and Léonie's [follow-up](https://tink.uk/perceived-affordances-and-the-functionality-mismatch/)), which is that we have to be careful about not mismatching user expectations. Having UI components that pretty much look the same (i.e. a row of square icon buttons with no enclosing border) that are sometimes mutually exclusive and sometimes not, sometimes de-selectable and sometimes not, seems like it would result in a lot of mismatched expectations within the app.

So from a **design standpoint**, I would advocate for a bit more clarity around these affordances before we go further. 

We have three possible combinations, so for example a consistent system could look like:

1. mutually exclusive, de-selectable → has transition animation, no border
2. mutually exclusive, not de-selectable → has transition animation, has border
3. not mutually exclusive (de-selectable by nature) → no transition animation, no border

It's probably not super important to visually afford mutual exclusivity because it's obvious after the first interaction, but having a visual differentiator for de-selectability seems important to me. Thoughts?

---

From the components maintenance standpoint, we will extract whatever design pattern that's decided on into the components package so it can be reused elsewhere in a consistent way. Ideally that componentization happens before this PR merges, but if it's urgent we can clean it up later 🙂

**Comment by noisysocks** [comment]
> I'll absolutely close #43993 for this one here which is a complete implementation since being able to un-select is essential to preserve theme styles as they are w/ no user intervention.

Forgot about your PR sorry! Props @draganescu ❤️

> If that means some controls (like this one) occupy a single row, then I think it's worth it. In the future we could consider making the Inspector wider to better accommodate multi-column layouts.

A single row sounds good to me. I think right now the buttons are 36px with a 2px margin but I can make them 40px to align fully with the inputs.

> So from a **design standpoint**, I would advocate for a bit more clarity around these affordances before we go further.
> 
> We have three possible combinations, so for example a consistent system could look like:
> 
> 1. mutually exclusive, de-selectable → has transition animation, no border
> 2. mutually exclusive, not de-selectable → has transition animation, has border
> 3. not mutually exclusive (de-selectable by nature) → no transition animation, no border

I love this and completely agree! 💯 

I'd add that only when there is a border (2) should the group be in a single tab stop.

> From the components maintenance standpoint, we will extract whatever design pattern that's decided on into the components package so it can be reused elsewhere in a consistent way. Ideally that componentization happens before this PR merges, but if it's urgent we can clean it up later 🙂

Happy to look at doing this in a components-first way. Were you thinking that there would be separate components for the three patterns above, or could all three patterns be handled by `ToggleGroupControl` using e.g. `optional` and `multiple` props?

**Comment by jasmussen** [comment]
Lots of good thoughts here. Happy to try the [suggested outline](https://github.com/WordPress/gutenberg/pull/44067#issuecomment-1244203285).

Outside of the components work, let's allow the 4-button group a whole row for now. I don't think we should touch any dimensions at this point, meaning 32x32 remains the toggle size so it matches up with these buttons:

Screenshot 2022-09-13 at 08 52 58

**Comment by mirka** [comment]
> Were you thinking that there would be separate components for the three patterns above, or could all three patterns be handled by `ToggleGroupControl` using e.g. `optional` and `multiple` props?

It's a tough call but I'm leaning toward handling everything in ToggleGroupControl with two extra props. I imagine it would be confusing for a dev to pick out the correct component to use if we split them 🤔

I'll start working on this during the coming days and see how it goes. Might have to split somehow if the internal logic feels untenable.

**Comment by noisysocks** [comment]
> Outside of the components work, let's allow the 4-button group a whole row for now. I don't think we should touch any dimensions at this point, meaning 32x32 remains the toggle size so it matches up with these buttons:

Cool yep. I've changed the button size to be 32px with 4px of top/bottom margin so that it's centred with the 40px inputs. I've also changed _Letter case_ to take up a full row.

Here's how it looks now:

| Block settings | Global styles |
| - | - |
| Screen Shot 2022-09-14 at 12 36 37 | Screen Shot 2022-09-14 at 12 36 51 |

> I'll start working on this during the coming days and see how it goes. Might have to split somehow if the internal logic feels untenable.

Legend, thanks! 🙌 I'm happy to help—ping me.

I think in that case then let's proceed with this PR as a smallish win (it's nice to have consistency between block settings and global styles) and then work on trying to consolidate everything around using `ToggleGroupControl`. We have quite a few places where `ToggleGroupControl` could be used if it were configurable enough, e.g. take a look at the block inspector for the Navigation block.

**Review by jasmussen — APPROVED** [review]
Nice work, looks good!

Screenshot 2022-09-14 at 08 36 49
Screenshot 2022-09-14 at 08 37 07

The _explicit none_ appears to work as intended. It's still unclear if the minus is the best longer term icon, but that should't block this from landing, as it's an important tool. Thank you for your patience on this one.

I'd love a code review, but for the behavior, let's try this.

File: packages/block-editor/src/components/text-transform-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-transform-control/index.js]: It might still be handy to allow passing additional props through here, or at least a custom `className` so we reuse the existing `.edit-site-typography-panel__full-width-control` style to make this control span the full width of the panel as mentioned in the discussion on this PR.

**Review by aaronrobertshaw — COMMENTED** [review]
It's great to see the ability to "toggle off" items in these controls return. 👌 

✅ Letter case appears in the Heading section, but no other "root" level sections
✅ Letter case appears in the Typography panel for blocks that support these controls
✅ Modifying Letter case works
✅ Controls reflect `none` selection when set via theme.json
✅ Storybook examples for text transform/decoration controls look & function correctly

❓ Letter case buttons are squashed (not square). Occurs in both block editor and Global Styles panels.

Overall this PR works pretty well for me. There was just a small styling issue that I think we need to tweak first before landing this. 

| Block Inspector Controls | Global Style - Block | Global Styles - Heading |
|---|---|---|
| Screen Shot 2022-09-15 at 1 24 45 pm | Screen Shot 2022-09-15 at 1 25 18 pm | Screen Shot 2022-09-15 at 1 25 03 pm |

When I view the TextTransformControl in both the site or block editor, the buttons are not square. They are in the Storybook examples. I believe this is due to the Letter case control not being set to span both columns and thus has its width constrained. 

To change that I believe we'd need to remove the `single-column` class from the typography support hook's use of the control. Also, the controls in the Global Styles panels appear to use the `.edit-site-typography-panel__full-width-control` class. I've left an inline comment on that below.

**Comment by jasmussen** [comment]
Just to echo Aaron's comments, this is looking good except for the squashed buttons:

Screenshot 2022-09-15 at 10 40 06
Screenshot 2022-09-15 at 10 40 17

Ideally the toggled state is 32x32 and if letter case can't fit in a single column, it's okay if it has to span two columns.

**Comment by noisysocks** [comment]
🤦‍♂️🤦‍♂️🤦‍♂️ I forgot to `git push` the changes described in https://github.com/WordPress/gutenberg/pull/44067#issuecomment-1246166362.

File: packages/block-editor/src/components/text-transform-control/index.js
Comment (noisysocks) [reply packages/block-editor/src/components/text-transform-control/index.js]: I'll add back `className`. We don't need it for that exact purpose though because `single-column` is added to the `ToolsPanelitem` in block settings and, for some reason, a wrapping div in global styles.

**Comment by noisysocks** [comment]
OK try this now. I rebased it to include https://github.com/WordPress/gutenberg/pull/44142, implemented @aaronrobertshaw's suggestion about adding a `className` prop, AND REMEMBERED TO `GIT PUSH` MY CHANGES.

**Comment by noisysocks** [comment]
@ockham reckon we could include this in 6.1? It isn't technically a "bug" (more "missing functionality") but it (along with https://github.com/WordPress/gutenberg/pull/44142) does make global styles feel less broken. Entirely up to you!

**Review by aaronrobertshaw — APPROVED** [review]
LGTM! This is testing well now 🚢

**Comment by ockham** [comment]
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: 853051bdfe

**Pull request merged**
